### PR TITLE
Added Locking Mechanism to Reminders Handler

### PR DESCRIPTION
**Pull request by b-quachtran**
**Proposed changes**:
- Added lock_store as an input to MessageProcessor so that handle_reminder can properly lock the conversation
- Updated handle_reminder to lock the conversation before retrieving the tracker store & executing
- Updated default_processor to pass in InMemoryLockStore as a param for test cases
- Resolves Issue #7887 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by b-quachtran** [comment]
@wochinge I added lock_store as an input parameter to MessageProcessor so that handle_reminder could lock the conversation, but not entirely sure if this is the best way to do it. Wdyt?

File: rasa/core/processor.py [hunk 1]
```diff
@@ -348,28 +351,31 @@ async def handle_reminder(
...
-            logger.debug(
-                f"Canceled reminder because it is outdated ({reminder_event})."
-            )
-        else:
-            intent = reminder_event.intent
-            entities = reminder_event.entities or {}
-            await self.trigger_external_user_uttered(
-                intent, entities, tracker, output_channel
-            )
+            if not tracker:
```
Comment (wochinge) [inline rasa/core/processor.py:359]: the tracker can't be `None` in this case

File: rasa/core/processor.py [hunk 2]
```diff
@@ -348,28 +351,31 @@ async def handle_reminder(
     ) -> None:
         """Handle a reminder that is triggered asynchronously."""
 
-        tracker = await self.get_tracker_with_session_start(sender_id, output_channel)
-
-        if not tracker:
-            logger.warning(
-                f"Failed to retrieve tracker for conversation ID '{sender_id}'."
+        async with self.lock_store.lock(sender_id):
```
Comment (wochinge) [inline rasa/core/processor.py:354]: can you please add a test to avoid regression?

**Review by wochinge — COMMENTED** [review]
Thanks for providing the fix! Will you also implement that for 2.x?

**Comment by b-quachtran** [comment]
@wochinge Yes, definitely. I'll create a PR for 2.X as well once this fix is good to go

File: tests/core/test_processor.py [hunk 3]
```diff
@@ -181,6 +181,38 @@ async def test_reminder_scheduled(
     assert t.events[-1] == ActionExecuted("action_listen")
 
 
+async def test_reminder_lock(
+    default_channel: CollectingOutputChannel, default_processor: MessageProcessor
+):
+    from io import StringIO
+
+    logger = logging.getLogger("rasa.core.lock_store")
```
Comment (wochinge) [inline tests/core/test_processor.py:189]: please use the caplog fixture instead (e.g. see [here](https://github.com/RasaHQ/rasa/blob/9913c1d53fd8a78b4564ebabb18a0bde9a939611/tests/core/test_tracker_stores.py#L274) on its usage)

File: rasa/core/agent.py [hunk 4]
```diff
@@ -888,6 +888,7 @@ def create_processor(
             self.policy_ensemble,
             self.domain,
             self.tracker_store,
+            self.lock_store,
```
Comment (wochinge) [inline rasa/core/agent.py:891]: not a fan of handing down intrastructure components to the `processor` but should in my opinion addressed separately (e.g. as part of https://github.com/RasaHQ/rasa/issues/5257)

**Review by wochinge — COMMENTED** [review]
did you also test this manually? If so, how?

File: rasa/core/agent.py (hunk 4 above)
Comment (b-quachtran) [reply rasa/core/agent.py:891]: I couldn't think of a better way to do this without requiring re-factoring. Is there a better approach to this than passing it into processor?

**Comment by b-quachtran** [comment]
@wochinge Yes, I tested manually using the ReminderBot with redis-server running locally and the SocketIO channel. Redis logs showed the lock correctly being acquired and deleted when the reminder was handled.

File: rasa/core/agent.py (hunk 4 above)
Comment (wochinge) [reply rasa/core/agent.py:891]: Also can't think of one for now. I think it's fine for now 👍🏻

File: changelog/7895.bugfix.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Fixed bug where the conversation does not lock before handling a reminder event.
```
Comment (wochinge) [inline changelog/7895.bugfix.md:1]: this has to be a `rst` changelog for `1.x`

**Review by wochinge — APPROVED** [review]

**Comment by wochinge** [comment]
Nice job 💯

**Pull request merged**